### PR TITLE
fix panic when corner case happen in RandomDelayIter

### PR DESCRIPTION
**Pull request by al8n**

### CODE REVIEW

File: mixnet/client/src/sender.rs [hunk 1]
```diff
@@ -133,21 +133,60 @@ where
...
+        self.remaining_time = self.remaining_time.saturating_sub(delay);
         self.remaining_delays -= 1;
 
         Some(delay)
     }
 }
+
+#[test]
+fn test_random_delay_iter_zero_total_time() {
+    const TOTAL_DELAYS: u64 = 3;
```
Comment (danielSanchezQ) [inline mixnet/client/src/sender.rs:176]: This const could be global shared if you use for all tests.

File: mixnet/client/src/sender.rs [hunk 2]
```diff
@@ -133,21 +133,60 @@ where
             return None;
         }
 
+        // corner case: 0 average delay
+        if self.avg_delay == 0 {
+            self.remaining_delays -= 1;
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:138]: We can set remaining delays at line 135 and avoid putting it in every branch

File: mixnet/client/src/sender.rs [hunk 3]
```diff
@@ -133,21 +133,60 @@ where
...
+        if self.remaining_time == 0 {
+            self.remaining_delays -= 1;
+            return Some(0);
+        }
+
         // Calculate bounds to avoid extreme values
         let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0)
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:156]: not sure I understand this formula

**Review by Zeegomo — CHANGES_REQUESTED** [review]

File: mixnet/client/src/sender.rs [hunk 4]
```diff
@@ -133,21 +133,60 @@ where
...
             as u64;
```
Comment (youngjoon-lee) [inline mixnet/client/src/sender.rs:157]: It would be helpful if we add a comment that explains why we need this `min(...)`. I understood it's for guaranteeing at least 1ms delay for all remaining hops. Did I understand it correctly?

File: mixnet/client/src/sender.rs [hunk 5]
```diff
@@ -133,21 +133,60 @@ where
...
 
+        // corner case: when lower bound is greater than upper bound
+        if lower_bound >= upper_bound {
+            self.remaining_delays -= 1;
+            self.remaining_time = 0;
+            return Some(self.remaining_time);
+        }
```
Comment (youngjoon-lee) [inline mixnet/client/src/sender.rs:164]: I'm thinking about the way to prevent this corner case, since this is not very clear to me :) but couldn't find the way yet.

File: mixnet/client/src/sender.rs (hunk 4 above)
Comment (al8n) [reply mixnet/client/src/sender.rs:157]: Sorry for missing the comment, and yes, this is a guarantee of at least a 1ms delay for all remaining hops.

File: mixnet/client/src/sender.rs (hunk 3 above)
Comment (al8n) [reply mixnet/client/src/sender.rs:156]: The line

```rust
.min((self.remaining_time - self.remaining_delays + 1) as f64)
```

is included to ensure that `upper_bound` never exceeds the maximum possible delay that can be returned without causing future delays to be zero or negative.

`self.remaining_time` is the total time still available for all remaining delays, and `self.remaining_delays` is the number of remaining delays, then the expression `(self.remaining_time - self.remaining_delays + 1)` calculates the maximum time we can allocate for the current delay without making future delays invalid (i.e., zero or negative).

For example, consider a case where we have `remaining_time = 100` and `remaining_delays = 10`. If we use up 95 time units for the current delay, then we would be left with only 5-time units to distribute among the remaining 9 delays, which is not possible without using zero or negative delays.

Here's how the logic works:

- `self.remaining_time` is the remaining time that needs to be allocated across all remaining delays.
- `self.remaining_delays` is the number of remaining delays.

The maximum you could use for the current delay to be still able to allocate at least 1 unit of time for each of the remaining delays is `(self.remaining_time - self.remaining_delays + 1)`.

The `min` function is used to choose the smaller of the two computed upper bounds (`1.5 * avg_delay` and `(self.remaining_time - self.remaining_delays + 1)`) to make sure that the program doesn't allocate too much time for the current delay.

Comment (Zeegomo) [reply mixnet/client/src/sender.rs:156]: Got it, better to put this in a comment as well (it's enought to point out that we don't want future delays to be zero)

File: mixnet/client/src/sender.rs [hunk 6]
```diff
@@ -133,60 +133,58 @@ where
...
-            self.remaining_time = 0;
-            return Some(self.remaining_time);
-        }
+            // guarantee that we don't exceed the remaining time and promise the delay we return is
+            // at least 1ms.
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0);
+        let lower_bound = (self.avg_delay as f64 * 0.5).min(upper_bound);
 
-        let delay = self.rng.gen_range(lower_bound, upper_bound);
+        let delay = self.rng.gen_range(lower_bound, upper_bound).floor() as u64;
```
[the remark this answers is not in the document]
Comment (Zeegomo) [reply mixnet/client/src/sender.rs:159]: well, what happens then if lower bound == upper bound? I think it probably panics

[the remark this answers is not in the document]
Comment (al8n) [reply mixnet/client/src/sender.rs:159]: Ah, yes, my bad, thanks!

File: mixnet/client/src/sender.rs [hunk 7]
```diff
@@ -133,21 +133,58 @@ where
...
-            as u64;
+            // guarantee that we don't exceed the remaining time and promise the delay we return is
+            // at least 1ms.
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0);
+        let lower_bound = (self.avg_delay as f64 * 0.5).min(upper_bound);
 
-        let delay = self.rng.gen_range(lower_bound, upper_bound);
-        self.remaining_time -= delay;
-        self.remaining_delays -= 1;
+        let delay = self.rng.gen_range(lower_bound, upper_bound + 1.0).floor() as u64;
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:159]: not a fan of +1, as you need to do some calculations to show that it's safe to use and it's easy to mess it up with future refactors. The next api would be better, but since we have to use an old version of rand, I suggest we use https://docs.rs/rand/0.7.3/rand/distributions/struct.Uniform.html#method.new_inclusive

Comment (al8n) [reply mixnet/client/src/sender.rs]: Good suggestion! Thanks!

Comment (al8n) [reply mixnet/client/src/sender.rs]: Good suggestion! Thanks!

File: mixnet/client/src/sender.rs [hunk 8]
```diff
@@ -133,21 +133,58 @@ where
...
+        }
+
         // Calculate bounds to avoid extreme values
-        let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
-            as u64;
+            // guarantee that we don't exceed the remaining time and promise the delay we return is
+            // at least 1ms.
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0);
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:156]: `.min((self.remaining_time as f64).saturating_sub(self.remaining_delays - 1))` and I think we can remove both the avg delay and remaining time edge cases

Comment (al8n) [reply mixnet/client/src/sender.rs:156]: If we remove the avg delay, then we may have extreme delays, and with higher possibility delay1 > delay 2 > delay3 and etc.

Comment (Zeegomo) [reply mixnet/client/src/sender.rs:156]: nono, I mean just removing the edge case at line 139

Comment (al8n) [reply mixnet/client/src/sender.rs:156]: Ah, yeah, sorry for the misunderstanding.

Comment (Zeegomo) [reply mixnet/client/src/sender.rs:146]: I still think we need the saturating sub in `.min((self.remaining_time as f64).saturating_sub(self.remaining_delays - 1))`

File: mixnet/client/src/sender.rs [hunk 9]
```diff
@@ -133,21 +133,48 @@ where
...
         }
 
         // Calculate bounds to avoid extreme values
-        let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
-            as u64;
+            // guarantee that we don't exceed the remaining time and promise the delay we return is
+            // at least 1ms.
+            .min(self.remaining_time.saturating_sub(self.remaining_delays) as f64 + 1.0);
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:146]: we don't need the +1 anymore since we decrease `remaining_delays` before this

File: mixnet/client/src/sender.rs [hunk 10]
```diff
@@ -133,21 +133,47 @@ where
...
+        assert!(delays.next().is_none());
+    }
+
+    #[test]
+    fn test_random_delay_iter_small_total_time() {
+        let mut delays = RandomDelayIterator::new(rand::thread_rng(), TOTAL_DELAYS, 1);
+        for _ in 0..TOTAL_DELAYS {
+            assert!(delays.next().is_some());
+        }
+        assert!(delays.next().is_none());
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:177]: we should probably check that the total wait is indeed 1, also a duration instead of a u64 would be much better because otherwise it's not clear what unit we are using

**Review by Zeegomo — CHANGES_REQUESTED** [review]

File: mixnet/client/src/sender.rs [hunk 11]
```diff
@@ -126,28 +126,58 @@ impl<R> Iterator for RandomDelayIterator<R>
...
+        let mut delays = RandomDelayIterator::new(rand::thread_rng(), TOTAL_DELAYS, 0);
+        for _ in 0..TOTAL_DELAYS {
+            assert!(delays.next().is_some());
+        }
+        assert!(delays.next().is_none());
+    }
+
+    #[test]
+    fn test_random_delay_iter_small_total_time() {
+        let mut delays = RandomDelayIterator::new(rand::thread_rng(), TOTAL_DELAYS, 1);
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:175]: in `RandomDelayIterator::new(rng, total_delays, total_time)`, `total_time` should be a `Duration` and not a `u64`

**Review by Zeegomo — CHANGES_REQUESTED** [review]

**Review by Zeegomo — APPROVED** [review]
